### ISSUE TITLE
[Troubleshoot]: Tried RECAP for lehome cloth folding challenge

### ISSUE THREAD

**Issue by Manu752** [RLinf/RLinf#1160]
https://github.com/user-attachments/assets/1f093824-2400-42c0-b403-1574e64201ea

<img width="1166" height="846" alt="Image" src="https://github.com/user-attachments/assets/2c878682-74c5-43e9-b8fd-a61da0806f9a" />
<img width="1166" height="846" alt="Image" src="https://github.com/user-attachments/assets/bf3970af-b72e-40d9-af06-b25bd87097a6" />

RECAP CFG Training: CFG-Trained Checkpoint (15k Steps) Performs Worse Than SFT Baseline (22k Steps) on LeHome Garment Folding
=================================================================================================================================

Environment
-----------
- Framework: RLinf RECAP pipeline
- Model: Pi0.5 (~3B params, HuggingFace/OpenPI format)
- Task: LeHome garment-folding benchmark (IsaacLab + SO-101 arm)
- Hardware: Single NVIDIA RTX PRO 6000 Blackwell (96 GB VRAM)
- Python: 3.11, PyTorch 2.7.1+cu128
- OpenPI config: pi05_so101_lehome

Background
----------
We followed the RECAP pipeline (Steps 1-5) on a garment-folding task. The SFT baseline
(checkpoint at step 22,000, trained on expert demonstrations only) performed well on the
evaluation. We then ran Step 4 (CFG training) for 15,000 steps on top of the 22k SFT
checkpoint. When we evaluated the CFG-trained checkpoint using proper two-pass guided
inference, it performed significantly worse than the SFT baseline -- the robot barely moved
the garment and never completed a fold.

Training Setup (Step 4: CFG Training)
--------------------------------------
Config file: examples/recap/cfg/config/lehome_cfg_openpi.yaml

Starting checkpoint: /openpi/checkpoints/pi05_so101_lehome/full_sft_v1/22000
(22k-step SFT checkpoint)

Model section:
  actor:
    model:
      model_type: "cfg_model"
      add_value_head: False
      openpi:
        config_name: "pi05_lehome"
        cfgrl_guidance_scale: 1.0
        unconditional_prob: 0.1
        train_expert_only: true
        guidance_type: "positive"
        positive_only_conditional: true

Optimizer:
  optim:
    lr: 5.0e-6
    adam_beta1: 0.9
    adam_beta2: 0.95
    adam_eps: 1.0e-8
    weight_decay: 1.0e-10
    clip_grad: 1.0
    lr_scheduler: "cosine"
    lr_warmup_steps: 500
    total_training_steps: 30000
    min_lr: 0.0

Training loop:
  runner:
    task_type: sft
    max_epochs: 15000
    max_steps: -1
    val_check_interval: -1
    save_interval: 5000

  actor:
    micro_batch_size: 4
    global_batch_size: 64

FSDP config:
  sharding_strategy: no_shard
  gradient_checkpointing: False
  skip_fsdp_wrapping: True
  mixed_precision: null

Data:
  data:
    advantage_tag: "fail800"
    balance_dataset_weights: true
    seed: 42
    num_workers: 4
    train_data_paths:
      - dataset_path: ".../top_long_merged"
        weight: 2.0
      - dataset_path: ".../lora_rollouts"
        weight: 1.0

Datasets
--------

Dataset 1: Expert SFT Demonstrations (top_long_merged)
  Episodes:                    250
  Total frames:                83,068
  Positive frames:             83,068 (100.0%)
  Negative frames:             0 (0%)
  Advantage source:            meta/advantages_fail800.parquet
  Advantage continuous range:  min=-0.1042, max=+0.1040, mean=0.0000
  Sampling weight:             2.0

  All 250 episodes are expert demonstrations (all frames labeled advantage=True).
  The advantage continuous values are tightly clustered near zero (range of only +-0.104),
  reflecting minimal variance in a dataset where all episodes succeed and the advantage
  signal is computed relative to a threshold of failure_reward = -800.

Dataset 2: RL Rollouts (lora_rollouts)
  Total episodes:              156
  Positive episodes:           156 (episodes that appear in positive frames)
  Negative episodes:           156 (episodes that appear in negative frames)
  Total frames:                85,719
  Positive frames:             29,481 (34.4%)
  Negative frames:             56,238 (65.6%)
  Advantage source:            meta/advantages_fail800.parquet
  Advantage continuous range:  min=-0.5521, max=+0.3693, mean=0.0000
  Sampling weight:             1.0

  The rollouts were collected after partial LoRA fine-tuning. The 156 episodes span a mix
  of successes and failures; 65.6% of frames are labeled negative (failure).

Effective Batch Composition (with balance_dataset_weights: true):
  With weight 2.0 on SFT and 1.0 on rollouts:
  - SFT contributes ~66% of effective samples (166,136 effective frames)
  - Rollouts contribute ~34% of effective samples (85,719 effective frames)

CFG Routing (with positive_only_conditional: true, unconditional_prob: 0.1):
  Route            | Prompt at training                          | Fraction | Content
  -----------------+---------------------------------------------+----------+-----------------------------------------------
  Conditional      | "fold the garment\nAdvantage: positive"     | ~70%     | 100% good actions (expert + success rollouts)
  Unconditional    | "fold the garment" (base prompt)            | ~30%     | ~26% good (dropout) + ~74% failure actions

So the unconditional (base prompt) branch is trained on a majority of failure actions.

Training Metrics (from TensorBoard)
-------------------------------------
Training ran for 15,000 steps. Checkpoints saved at steps 5,000, 10,000, and 15,000.

Overall Loss (train/loss):
  Step 0:      0.08192
  Step 499:    0.04764
  Step 1,000:  0.04332
  Step 2,001:  0.03775
  Step 4,999:  0.03336
  Step 7,502:  0.03496
  Step 10,000: 0.02865
  Step 12,500: 0.02866
  Step 15,000: 0.02500

Conditional Loss (train/conditional_loss):
  Step 0:      0.07836
  Step 499:    0.04101
  Step 2,001:  0.03555
  Step 4,999:  0.03106
  Step 7,502:  0.02777
  Step 10,000: 0.03071
  Step 15,000: 0.02356

Unconditional Loss (train/unconditional_loss):
  Step 0:      0.09035
  Step 499:    0.05798
  Step 2,001:  0.04295
  Step 4,999:  0.03883
  Step 7,502:  0.04617
  Step 10,000: 0.02445
  Step 15,000: 0.03196

Positive Conditional Loss (train/positive_conditional_loss):
  (Identical to conditional_loss since positive_only_conditional: true means
  only positive samples ever reach the conditional branch)
  Step 0:      0.07836
  Step 4,999:  0.03106
  Step 15,000: 0.02356

Positive Unconditional Loss (train/positive_unconditional_loss):
  (Good actions routed to unconditional branch due to 10% dropout)
  Step 0:      0.09004
  Step 999:    0.05560
  Step 2,002:  0.03289
  Step 5,001:  0.02567
  Step 7,500:  0.02240
  Step 10,001: 0.03581
  Step 12,500: 0.02113
  Step 15,000: 0.00720

Negative Unconditional Loss (train/negative_unconditional_loss):
  (Failure actions always routed to unconditional branch)
  Step 0:      0.09049
  Step 499:    0.05831
  Step 2,001:  0.04788
  Step 4,999:  0.03934
  Step 7,502:  0.05439
  Step 10,000: 0.02506
  Step 12,500: 0.02969
  Step 15,000: 0.03443

Gradient Norm (train/grad_norm):
  Step 0:      0.233
  Step 499:    0.092
  Step 1,000:  0.085
  Step 2,001:  0.086
  Step 4,999:  0.104
  Step 7,502:  0.112
  Step 10,000: 0.124
  Step 12,500: 0.163
  Step 15,000: 0.130

  Grad norm stabilized around 0.09-0.13 throughout training with no spikes.
  The overall loss decreased monotonically.

Learning Rate (train/learning_rate):
  Cosine schedule with 500 warmup steps, peak 5e-6:
  Step 0:      0.0 (warmup start)
  Step 100:    9.99e-7
  Step 250:    2.50e-6
  Step 499:    4.99e-6 (peak)
  Step 750:    4.999e-6
  Step 1,000:  4.996e-6
  Step 2,001:  4.968e-6
  Step 5,000:  4.719e-6
  Step 10,000: 3.826e-6
  Step 15,000: 2.567e-6

Label and Routing Ratios (per-batch, sampled across training):
  train/positive_label_ratio:         ranged ~0.66-0.84
  train/negative_label_ratio:         ranged ~0.16-0.34
  train/conditional_ratio:            ranged ~0.61-0.83
  train/unconditional_ratio:          ranged ~0.17-0.39

Evaluation Setup
----------------
Inference script: Custom two-pass CFG server (serve_policy_cfg.py) written to properly
implement guided inference, since the standard serve_policy.py does only a single
unconditional denoising pass. Using serve_policy.py with a CFG-trained checkpoint means
the model is evaluated as if it is unconditional -- the conditional branch is never
activated and the guidance is never applied.

CFG inference formula:
  v_t = (1 - w) * v_uncond + w * v_cond

Two-pass computation per denoising step:
  1. Unconditional pass: base prompt "fold the garment"
     -> KV cache pre-computed -> v_uncond
  2. Conditional pass: "fold the garment\nAdvantage: positive"
     -> separate KV cache pre-computed -> v_cond
  Combined: v_t = (1 - 1.5) * v_uncond + 1.5 * v_cond

Server parameters used during eval:
  --config          pi05_so101_lehome
  --checkpoint-dir  .../global_step_15000/serve
  --guidance-scale  1.5
  --num-steps       10
  --port            8000
  --device          cuda

Evaluation command:
  python -m scripts.eval \
    --policy_type openpi \
    --openpi_host localhost \
    --openpi_port 8000 \
    --openpi_action_chunk_size 50 \
    --openpi_replan_steps 5 \
    --num_episodes 5

IsaacLab simulation: --device cpu
(required to avoid GPU contention with the policy server on the single 96GB GPU)

Observed Behavior
-----------------
- The robot arm moves but produces only very small, hesitant motions
- Garment fold distances recorded per episode: approximately 25-47
  (task success threshold is <= 7-9)
- The distances do not meaningfully decrease across timesteps -- the robot is not
  progressively folding the garment
- All 5 evaluated episodes failed
- The original 22k SFT checkpoint (served with standard single-pass inference using
  the base prompt "fold the garment") reliably folds the garment

Comparison
----------
Setup            | Checkpoint        | Inference type   | Guidance scale | Observed behavior
-----------------+-------------------+------------------+----------------+---------------------------
SFT baseline     | 22k SFT           | Single-pass uncond| N/A           | Successfully folds garment
CFG trained      | 15k CFG on 22k    | Two-pass CFG     | 1.5            | Robot barely moves, all fail

Please Assist :)

**Comment by Nadeem752** [comment]
@QuanluZhang  could you please assist ?

**Comment by zoeyuchao** [comment]
Hi, we found RECAP does have some limitations. So RLinf team are working on improving it and will release paper in the next two weeks.

**Comment by Nadeem752** [comment]
@zoeyuchao  has the paper or  the code been released ?
